Failed to merge directives

I am using entgo and the framework generate and apply directives on the schema file. The gqlmerge is failing to merge schemas with these directives.

schema.graphql
```
directive @goModel(
  model: String
  models: [String!]
) on OBJECT | INPUT_OBJECT | SCALAR | ENUM | INTERFACE | UNION

enum Color @goModel(model: "backend/ent/color.Color") {
  Blue
  Red
}
```

Error
```
🎉 [graph/auth] Total 6 GraphQL files found!
panic: /Users/x/Documents/Projects/www/backend/graph/auth/register.mutation.graphql:1:11: unexpected "@", expected Ident

goroutine 1 [running]:
github.com/mattdamon108/gqlmerge/lib.(*Lexer).ConsumeToken(0x14000100280, 0xfffffffe)
        /Users/x/go/pkg/mod/github.com/mattdamon108/gqlmerge@v0.2.5/lib/lexer.go:104 +0x16c
github.com/mattdamon108/gqlmerge/lib.(*Lexer).ConsumeIdent(0x14000100280)
        /Users/x/go/pkg/mod/github.com/mattdamon108/gqlmerge@v0.2.5/lib/lexer.go:88 +0x3c
github.com/mattdamon108/gqlmerge/lib.(*Schema).ParseSchema(0x140001060f0, 0x14000100280)
        /Users/x/go/pkg/mod/github.com/mattdamon108/gqlmerge@v0.2.5/lib/parse.go:15 +0x44
github.com/mattdamon108/gqlmerge/lib.getSchema({0x16d53b7ce?, 0xa?})
        /Users/x/go/pkg/mod/github.com/mattdamon108/gqlmerge@v0.2.5/lib/merge.go:50 +0xd0
github.com/mattdamon108/gqlmerge/lib.Merge({0x1400001406c, 0x4}, {0x140000101c0, 0x1, 0x1400004a6b8?})
        /Users/x/go/pkg/mod/github.com/mattdamon108/gqlmerge@v0.2.5/lib/merge.go:18 +0xa0
main.main()
        /Users/x/go/pkg/mod/github.com/mattdamon108/gqlmerge@v0.2.5/main.go:21 +0xbc
exit status 2
```

--- a/.gitignore
+++ b/.gitignore
@@ -1,4 +1,3 @@
-schema/
 gqlmerge
 
 *.gql
@@ -9,3 +8,4 @@ gqlmerge
 .idea
 
 demo/demo.yml
+!test/*.graphql

--- a/README.md
+++ b/README.md
@@ -25,19 +25,19 @@ A tool to merge and stitch modularized GraphQL files into one schema file
 Homebrew
 
 ```shell
-$ brew install mattdamon108/tools/gqlmerge
+$ brew install mununki/tools/gqlmerge
 ```
 
 Using `go get`
 
 ```shell
-$ go get -u github.com/mattdamon108/gqlmerge
+$ go get -u github.com/mununki/gqlmerge
 ```
 
 Building with source code
 
 ```shell
-$ git clone https://github.com/mattdamon108/gqlmerge
+$ git clone https://github.com/mununki/gqlmerge
 
 $ cd gqlmerge
 
@@ -49,7 +49,7 @@ $ go install
 Import gqlmerge module
 
 ```go
-import gql "github.com/mattdamon108/gqlmerge/lib"
+import gql "github.com/mununki/gqlmerge/lib"
 
 func main(){
 	// ...

--- a/go.mod
+++ b/go.mod
@@ -1,3 +1,3 @@
-module github.com/mattdamon108/gqlmerge
+module github.com/mununki/gqlmerge
 
 go 1.12

--- a/lib/lexer.go
+++ b/lib/lexer.go
@@ -29,25 +29,18 @@ func (l *Lexer) ConsumeWhitespace() {
 	for {
 		l.next = l.sc.Scan()
 
-		// if l.next == ',' {
-		// 	continue
-		// }
+		if l.next == '#' {
+			l.ConsumeComment()
+			continue
+		}
 
-		// if l.next == '#' {
-		// 	l.ConsumeComment()
-		// 	continue
-		// }
-
-		// // GraphQL comments """ ... """
-		// // due to scanner recognize the GraphQL comments as string
-		// // (first loop) ""
-		// // (second loop) " ... "
-		// // (third loop) ""
-		// // finally buffer gets """ ... """
-		// if l.next == scanner.String {
-		// 	l.ConsumeMultiLineComment()
-		// 	continue
-		// }
+		if l.next == scanner.String {
+			if l.sc.Peek() != '"' {
+				break
+			}
+			l.ConsumeMultiLineComment()
+			continue
+		}
 
 		break
 	}
@@ -64,16 +57,16 @@ func (l *Lexer) ConsumeComment() {
 }
 
 func (l *Lexer) ConsumeMultiLineComment() {
-	next := l.sc.TokenText()
-	l.buffer.WriteString(next)
-}
-
-func (l *Lexer) ConsumeDirective() {
-	l.buffer.WriteString(l.sc.TokenText())
 	for {
-		next := l.sc.Next()
-		if next == '\r' || next == '\n' || next == scanner.EOF {
+		next := l.sc.Scan()
+		if next == scanner.EOF {
 			break
+		}
+		if next == scanner.String {
+			next := l.sc.Next()
+			if next == '"' {
+				break
+			}
 		}
 		l.buffer.WriteRune(next)
 	}

--- a/lib/lexer.go
+++ b/lib/lexer.go
@@ -29,25 +29,25 @@ func (l *Lexer) ConsumeWhitespace() {
 	for {
 		l.next = l.sc.Scan()
 
-		if l.next == ',' {
-			continue
-		}
+		// if l.next == ',' {
+		// 	continue
+		// }
 
-		if l.next == '#' {
-			l.ConsumeComment()
-			continue
-		}
+		// if l.next == '#' {
+		// 	l.ConsumeComment()
+		// 	continue
+		// }
 
-		// GraphQL comments """ ... """
-		// due to scanner recognize the GraphQL comments as string
-		// (first loop) ""
-		// (second loop) " ... "
-		// (third loop) ""
-		// finally buffer gets """ ... """
-		if l.next == scanner.String {
-			l.ConsumeMultiLineComment()
-			continue
-		}
+		// // GraphQL comments """ ... """
+		// // due to scanner recognize the GraphQL comments as string
+		// // (first loop) ""
+		// // (second loop) " ... "
+		// // (third loop) ""
+		// // finally buffer gets """ ... """
+		// if l.next == scanner.String {
+		// 	l.ConsumeMultiLineComment()
+		// 	continue
+		// }
 
 		break
 	}
@@ -87,6 +87,12 @@ func (l *Lexer) ConsumeIdent() string {
 	name := l.sc.TokenText()
 	l.ConsumeToken(scanner.Ident)
 	return name
+}
+
+func (l *Lexer) ConsumeString() string {
+	str := l.sc.TokenText()
+	l.ConsumeWhitespace()
+	return str
 }
 
 func (l *Lexer) ConsumeToken(expected rune) {

--- a/lib/merge.go
+++ b/lib/merge.go
@@ -58,6 +58,7 @@ func joinSchemas(schemas []Schema) *Schema {
 
 	for _, s := range schemas {
 		schema.Files = append(schema.Files, s.Files...)
+		schema.DirectiveDefinitions = append(schema.DirectiveDefinitions, s.DirectiveDefinitions...)
 		schema.Mutations = append(schema.Mutations, s.Mutations...)
 		schema.Queries = append(schema.Queries, s.Queries...)
 		schema.Subscriptions = append(schema.Subscriptions, s.Subscriptions...)
@@ -70,8 +71,9 @@ func joinSchemas(schemas []Schema) *Schema {
 	}
 
 	wg := sync.WaitGroup{}
-	wg.Add(9)
+	wg.Add(10)
 
+	go schema.UniqueDirectiveDefinition(&wg)
 	go schema.UniqueMutation(&wg)
 	go schema.UniqueQuery(&wg)
 	go schema.UniqueSubscription(&wg)

--- a/lib/parse.go
+++ b/lib/parse.go
@@ -26,12 +26,39 @@ func (s *Schema) ParseSchema(l *Lexer) {
 			}
 			l.ConsumeToken('}')
 
+		case "directive":
+			d := DirectiveDefinition{}
+			d.Filename = l.sc.Filename
+			d.Line = l.sc.Line
+			d.Column = l.sc.Column
+			l.ConsumeToken('@')
+			d.Name = l.ConsumeIdent()
+			d.Args = ParseArgument(l)
+			if l.ConsumeIdent() == "repeatable" {
+				d.Repeatable = true
+			} else {
+				// "on" should be consumed already here
+				d.Repeatable = false
+				ls := []string{}
+				for l.Peek() != '\r' || l.Peek() != '\n' || l.Peek() != scanner.EOF {
+					ls = append(ls, l.ConsumeIdent())
+					if l.Peek() == '|' {
+						l.ConsumeToken('|')
+					} else {
+						break
+					}
+				}
+				d.Locations = ls
+			}
+			s.DirectiveDefinitions = append(s.DirectiveDefinitions, &d)
+
 		case "scalar":
 			c := Scalar{}
 			c.Filename = l.sc.Filename
 			c.Line = l.sc.Line
 			c.Column = l.sc.Column
 			c.Name = l.ConsumeIdent()
+			c.Directives = ParseDirectives(l)
 			s.Scalars = append(s.Scalars, &c)
 
 		case "enum":
@@ -40,9 +67,13 @@ func (s *Schema) ParseSchema(l *Lexer) {
 			e.Line = l.sc.Line
 			e.Column = l.sc.Column
 			e.Name = l.ConsumeIdent()
+			e.Directives = ParseDirectives(l)
 			l.ConsumeToken('{')
 			for l.Peek() != '}' {
-				e.Fields = append(e.Fields, l.ConsumeIdent())
+				ev := EnumValue{}
+				ev.Name = l.ConsumeIdent()
+				ev.Directives = ParseDirectives(l)
+				e.EnumValues = append(e.EnumValues, ev)
 			}
 			l.ConsumeToken('}')
 			s.Enums = append(s.Enums, &e)
@@ -53,6 +84,7 @@ func (s *Schema) ParseSchema(l *Lexer) {
 			i.Line = l.sc.Line
 			i.Column = l.sc.Column
 			i.Name = l.ConsumeIdent()
+			i.Directives = ParseDirectives(l)
 
 			l.ConsumeToken('{')
 
@@ -94,11 +126,7 @@ func (s *Schema) ParseSchema(l *Lexer) {
 				}
 
 				if l.Peek() == '@' {
-					l.ConsumeToken('@')
-					l.ConsumeDirective()
-					d := Directive{l.GetBuffer()}
-					l.ConsumeWhitespace()
-					p.Directive = &d
+					p.Directives = ParseDirectives(l)
 				}
 
 				i.Props = append(i.Props, &p)
@@ -113,6 +141,7 @@ func (s *Schema) ParseSchema(l *Lexer) {
 			u.Line = l.sc.Line
 			u.Column = l.sc.Column
 			u.Name = l.ConsumeIdent()
+			u.Directives = ParseDirectives(l)
 			l.ConsumeToken('=')
 			for l.Peek() != '\r' || l.Peek() != '\n' || l.Peek() != scanner.EOF {
 				u.Fields = append(u.Fields, l.ConsumeIdent())
@@ -167,13 +196,7 @@ func (s *Schema) ParseSchema(l *Lexer) {
 					}
 				}
 
-				if l.Peek() == '@' {
-					l.ConsumeToken('@')
-					l.ConsumeDirective()
-					d := Directive{l.GetBuffer()}
-					l.ConsumeWhitespace()
-					p.Directive = &d
-				}
+				p.Directives = ParseDirectives(l)
 
 				i.Props = append(i.Props, &p)
 			}
@@ -229,13 +252,7 @@ func (s *Schema) ParseSchema(l *Lexer) {
 					}
 					q.Resp = r
 
-					if l.Peek() == '@' {
-						l.ConsumeToken('@')
-						l.ConsumeDirective()
-						d := Directive{l.GetBuffer()}
-						l.ConsumeWhitespace()
-						q.Directive = &d
-					}
+					q.Directives = ParseDirectives(l)
 
 					s.Queries = append(s.Queries, &q)
 				}
@@ -286,13 +303,7 @@ func (s *Schema) ParseSchema(l *Lexer) {
 
 					m.Resp = r
 
-					if l.Peek() == '@' {
-						l.ConsumeToken('@')
-						l.ConsumeDirective()
-						d := Directive{l.GetBuffer()}
-						l.ConsumeWhitespace()
-						m.Directive = &d
-					}
+					m.Directives = ParseDirectives(l)
 
 					s.Mutations = append(s.Mutations, &m)
 				}
@@ -342,13 +353,7 @@ func (s *Schema) ParseSchema(l *Lexer) {
 					}
 					c.Resp = r
 
-					if l.Peek() == '@' {
-						l.ConsumeToken('@')
-						l.ConsumeDirective()
-						d := Directive{l.GetBuffer()}
-						l.ConsumeWhitespace()
-						c.Directive = &d
-					}
+					c.Directives = ParseDirectives(l)
 
 					s.Subscriptions = append(s.Subscriptions, &c)
 				}
@@ -416,13 +421,7 @@ func (s *Schema) ParseSchema(l *Lexer) {
 						}
 					}
 
-					if l.Peek() == '@' {
-						l.ConsumeToken('@')
-						l.ConsumeDirective()
-						d := Directive{l.GetBuffer()}
-						l.ConsumeWhitespace()
-						p.Directive = &d
-					}
+					p.Directives = ParseDirectives(l)
 
 					t.Props = append(t.Props, &p)
 				}
@@ -434,6 +433,39 @@ func (s *Schema) ParseSchema(l *Lexer) {
 	}
 }
 
+func (s *Schema) UniqueDirectiveDefinition(wg *sync.WaitGroup) {
+	defer wg.Done()
+	j := 0
+	seen := make(map[string]struct{}, len(s.DirectiveDefinitions))
+	for _, v := range s.DirectiveDefinitions {
+		if _, ok := seen[v.Name]; ok {
+			for i := 0; i < j; i++ {
+				if s.DirectiveDefinitions[i].Name == v.Name {
+					if reflect.DeepEqual(s.DirectiveDefinitions[i].Args, v.Args) && reflect.DeepEqual(s.DirectiveDefinitions[i].Repeatable, v.Repeatable) && reflect.DeepEqual(s.DirectiveDefinitions[i].Locations, v.Locations) {
+						break
+					} else {
+						rel1, err := GetRelPath(s.Mutations[i].Filename)
+						if err != nil {
+							panic(err)
+						}
+						rel2, err := GetRelPath(v.Filename)
+						if err != nil {
+							panic(err)
+						}
+
+						panic(fmt.Sprintf("Duplicated Directive Definitions: %s(%s, %v:%v) and (%s, %v:%v)", s.DirectiveDefinitions[i].Name, *rel1, s.DirectiveDefinitions[i].Line, s.DirectiveDefinitions[i].Column, *rel2, v.Line, v.Column))
+					}
+				}
+			}
+			continue
+		}
+		seen[v.Name] = struct{}{}
+		s.DirectiveDefinitions[j] = v
+		j++
+	}
+	s.DirectiveDefinitions = s.DirectiveDefinitions[:j]
+}
+
 func (s *Schema) UniqueMutation(wg *sync.WaitGroup) {
 	defer wg.Done()
 	j := 0
@@ -442,7 +474,7 @@ func (s *Schema) UniqueMutation(wg *sync.WaitGroup) {
 		if _, ok := seen[v.Name]; ok {
 			for i := 0; i < j; i++ {
 				if s.Mutations[i].Name == v.Name {
-					if reflect.DeepEqual(s.Mutations[i].Args, v.Args) && reflect.DeepEqual(s.Mutations[i].Resp, v.Resp) && reflect.DeepEqual(s.Mutations[i].Directive, v.Directive) {
+					if reflect.DeepEqual(s.Mutations[i].Args, v.Args) && reflect.DeepEqual(s.Mutations[i].Resp, v.Resp) && reflect.DeepEqual(s.Mutations[i].Directives, v.Directives) {
 						break
 					} else {
 
@@ -476,7 +508,7 @@ func (s *Schema) UniqueQuery(wg *sync.WaitGroup) {
 		if _, ok := seen[v.Name]; ok {
 			for i := 0; i < j; i++ {
 				if s.Queries[i].Name == v.Name {
-					if reflect.DeepEqual(s.Queries[i].Args, v.Args) && reflect.DeepEqual(s.Queries[i].Resp, v.Resp) && reflect.DeepEqual(s.Queries[i].Directive, v.Directive) {
+					if reflect.DeepEqual(s.Queries[i].Args, v.Args) && reflect.DeepEqual(s.Queries[i].Resp, v.Resp) && reflect.DeepEqual(s.Queries[i].Directives, v.Directives) {
 						break
 					} else {
 
@@ -510,7 +542,7 @@ func (s *Schema) UniqueSubscription(wg *sync.WaitGroup) {
 		if _, ok := seen[v.Name]; ok {
 			for i := 0; i < j; i++ {
 				if s.Subscriptions[i].Name == v.Name {
-					if reflect.DeepEqual(s.Subscriptions[i].Args, v.Args) && reflect.DeepEqual(s.Subscriptions[i].Resp, v.Resp) && reflect.DeepEqual(s.Subscriptions[i].Directive, v.Directive) {
+					if reflect.DeepEqual(s.Subscriptions[i].Args, v.Args) && reflect.DeepEqual(s.Subscriptions[i].Resp, v.Resp) && reflect.DeepEqual(s.Subscriptions[i].Directives, v.Directives) {
 						break
 					} else {
 
@@ -544,7 +576,7 @@ func (s *Schema) UniqueTypeName(wg *sync.WaitGroup) {
 		if _, ok := seen[v.Name]; ok {
 			for i := 0; i < j; i++ {
 				if s.TypeNames[i].Name == v.Name {
-					if reflect.DeepEqual(s.TypeNames[i].ImplTypes, v.ImplTypes) && reflect.DeepEqual(s.TypeNames[i].Props, v.Props) {
+					if reflect.DeepEqual(s.TypeNames[i].ImplTypes, v.ImplTypes) && reflect.DeepEqual(s.TypeNames[i].Props, v.Props) && reflect.DeepEqual(s.TypeNames[i].Directives, v.Directives) {
 						break
 					} else {
 
@@ -578,18 +610,20 @@ func (s *Schema) UniqueScalar(wg *sync.WaitGroup) {
 		if _, ok := seen[v.Name]; ok {
 			for i := 0; i < j; i++ {
 				if s.Scalars[i].Name == v.Name {
-					break
-				} else {
-					rel1, err := GetRelPath(s.Scalars[i].Filename)
-					if err != nil {
-						panic(err)
-					}
-					rel2, err := GetRelPath(v.Filename)
-					if err != nil {
-						panic(err)
-					}
+					if reflect.DeepEqual(s.Scalars[i].Directives, v.Directives) {
+						break
+					} else {
+						rel1, err := GetRelPath(s.Scalars[i].Filename)
+						if err != nil {
+							panic(err)
+						}
+						rel2, err := GetRelPath(v.Filename)
+						if err != nil {
+							panic(err)
+						}
 
-					panic(fmt.Sprintf("Duplicated Scalars: %s(%s, %v:%v) and (%s, %v:%v)", s.Scalars[i].Name, *rel1, s.Scalars[i].Line, s.Scalars[i].Column, *rel2, v.Line, v.Column))
+						panic(fmt.Sprintf("Duplicated Scalars: %s(%s, %v:%v) and (%s, %v:%v)", s.Scalars[i].Name, *rel1, s.Scalars[i].Line, s.Scalars[i].Column, *rel2, v.Line, v.Column))
+					}
 				}
 			}
 			continue
@@ -609,7 +643,7 @@ func (s *Schema) UniqueEnum(wg *sync.WaitGroup) {
 		if _, ok := seen[v.Name]; ok {
 			for i := 0; i < j; i++ {
 				if s.Enums[i].Name == v.Name {
-					if reflect.DeepEqual(s.Enums[i].Fields, v.Fields) {
+					if reflect.DeepEqual(s.Enums[i].Directives, v.Directives) && reflect.DeepEqual(s.Enums[i].EnumValues, v.EnumValues) {
 						break
 					} else {
 
@@ -643,7 +677,7 @@ func (s *Schema) UniqueInterface(wg *sync.WaitGroup) {
 		if _, ok := seen[v.Name]; ok {
 			for i := 0; i < j; i++ {
 				if s.Interfaces[i].Name == v.Name {
-					if reflect.DeepEqual(s.Interfaces[i].Props, v.Props) {
+					if reflect.DeepEqual(s.Interfaces[i].Directives, v.Directives) && reflect.DeepEqual(s.Interfaces[i].Props, v.Props) {
 						break
 					} else {
 
@@ -677,7 +711,7 @@ func (s *Schema) UniqueUnion(wg *sync.WaitGroup) {
 		if _, ok := seen[v.Name]; ok {
 			for i := 0; i < j; i++ {
 				if s.Unions[i].Name == v.Name {
-					if reflect.DeepEqual(s.Unions[i].Fields, v.Fields) {
+					if reflect.DeepEqual(s.Unions[i].Directives, v.Directives) && reflect.DeepEqual(s.Unions[i].Fields, v.Fields) {
 						break
 					} else {
 
@@ -781,10 +815,63 @@ func ParseArgument(l *Lexer) []*Arg {
 					arg.Null = true
 				}
 			}
+			if l.Peek() == '@' {
+				arg.Directives = ParseDirectives(l)
+			}
 
 			args = append(args, &arg)
 		}
 		l.ConsumeToken(')')
 	}
 	return args
+}
+
+func ParseDirectives(l *Lexer) []*Directive {
+	ds := []*Directive{}
+
+	for l.Peek() == '@' {
+		l.ConsumeToken('@')
+		d := Directive{}
+		d.Name = l.ConsumeIdent()
+
+		for l.Peek() == '(' {
+			l.ConsumeToken('(')
+			for l.Peek() != ')' {
+				da := DirectiveArg{}
+				da.Name = l.ConsumeIdent()
+				l.ConsumeToken(':')
+
+				if l.Peek() == '[' {
+					da.IsList = true
+					l.ConsumeToken('[')
+					da.Value = ParseList(l)
+					l.ConsumeToken(']')
+				} else {
+					da.IsList = false
+					if l.Peek() == scanner.String {
+						da.Value = append(da.Value, l.ConsumeString())
+					} else {
+						da.Value = append(da.Value, l.ConsumeIdent())
+					}
+				}
+
+				d.DirectiveArgs = append(d.DirectiveArgs, &da)
+			}
+			l.ConsumeToken(')')
+		}
+		ds = append(ds, &d)
+	}
+	return ds
+}
+
+func ParseList(l *Lexer) []string {
+	ss := []string{}
+	for l.Peek() != ']' {
+		if l.Peek() == scanner.String {
+			ss = append(ss, l.ConsumeString())
+		} else {
+			ss = append(ss, l.ConsumeIdent())
+		}
+	}
+	return ss
 }

--- a/lib/parse_string_test.go
+++ b/lib/parse_string_test.go
@@ -1,0 +1,21 @@
+package lib
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"text/scanner"
+)
+
+func TestParseString(t *testing.T) {
+	const src = `(model: "todo/ent.Noder")`
+
+	var s scanner.Scanner
+	s.Init(strings.NewReader(src))
+
+	for tok := s.Scan(); tok != scanner.EOF; tok = s.Scan() {
+		if tok == scanner.String {
+			fmt.Printf("Found a string: %s\n", s.TokenText())
+		}
+	}
+}

--- a/lib/parse_string_test.go
+++ b/lib/parse_string_test.go
@@ -8,14 +8,24 @@ import (
 )
 
 func TestParseString(t *testing.T) {
-	const src = `(model: "todo/ent.Noder")`
+	const src = `"""
+TEST
+"""`
 
 	var s scanner.Scanner
 	s.Init(strings.NewReader(src))
 
-	for tok := s.Scan(); tok != scanner.EOF; tok = s.Scan() {
-		if tok == scanner.String {
-			fmt.Printf("Found a string: %s\n", s.TokenText())
+	for {
+		next := s.Scan()
+		fmt.Println(s.Pos(), s.Position.Offset)
+		if next == scanner.String {
+			next = s.Next()
+			if next == '"' {
+				fmt.Printf("Found a string: %s\n", s.TokenText())
+			}
+		}
+		if next == scanner.EOF {
+			break
 		}
 	}
 }

--- a/lib/types.go
+++ b/lib/types.go
@@ -11,49 +11,51 @@ type BaseFileInfo struct {
 }
 
 type Schema struct {
-	Files         []*os.File
-	Mutations     []*Mutation
-	Queries       []*Query
-	Subscriptions []*Subscription
-	TypeNames     []*TypeName
-	Scalars       []*Scalar
-	Enums         []*Enum
-	Interfaces    []*Interface
-	Unions        []*Union
-	Inputs        []*Input
+	Files                []*os.File
+	Mutations            []*Mutation
+	Queries              []*Query
+	Subscriptions        []*Subscription
+	TypeNames            []*TypeName
+	Scalars              []*Scalar
+	Enums                []*Enum
+	Interfaces           []*Interface
+	Unions               []*Union
+	Inputs               []*Input
+	DirectiveDefinitions []*DirectiveDefinition
 }
 
 type Mutation struct {
 	BaseFileInfo
-	Name      string
-	Args      []*Arg
-	Resp      Resp
-	Directive *Directive
+	Name       string
+	Args       []*Arg
+	Resp       Resp
+	Directives []*Directive
 }
 
 type Query struct {
 	BaseFileInfo
-	Name      string
-	Args      []*Arg
-	Resp      Resp
-	Directive *Directive
+	Name       string
+	Args       []*Arg
+	Resp       Resp
+	Directives []*Directive
 }
 
 type Subscription struct {
 	BaseFileInfo
-	Name      string
-	Args      []*Arg
-	Resp      Resp
-	Directive *Directive
+	Name       string
+	Args       []*Arg
+	Resp       Resp
+	Directives []*Directive
 }
 
 type TypeName struct {
 	BaseFileInfo
-	Name      string
-	Impl      bool
-	ImplType  *string // deprecated, use ImplTypes
-	ImplTypes []string
-	Props     []*Prop
+	Name       string
+	Impl       bool
+	ImplType   *string // deprecated, use ImplTypes
+	ImplTypes  []string
+	Props      []*Prop
+	Directives []*Directive
 }
 
 type Arg struct {
@@ -63,6 +65,7 @@ type Arg struct {
 	Null       bool
 	IsList     bool
 	IsListNull bool
+	Directives []*Directive
 }
 
 type Resp struct {
@@ -79,30 +82,39 @@ type Prop struct {
 	Null       bool
 	IsList     bool
 	IsListNull bool
-	Directive  *Directive
+	Directives []*Directive
 }
 
 type Scalar struct {
 	BaseFileInfo
-	Name string
+	Name       string
+	Directives []*Directive
 }
 
 type Enum struct {
 	BaseFileInfo
-	Name   string
-	Fields []string
+	Name       string
+	EnumValues []EnumValue
+	Directives []*Directive
+}
+
+type EnumValue struct {
+	Name       string
+	Directives []*Directive
 }
 
 type Interface struct {
 	BaseFileInfo
-	Name  string
-	Props []*Prop
+	Name       string
+	Props      []*Prop
+	Directives []*Directive
 }
 
 type Union struct {
 	BaseFileInfo
-	Name   string
-	Fields []string
+	Name       string
+	Fields     []string
+	Directives []*Directive
 }
 
 type Input struct {
@@ -111,6 +123,21 @@ type Input struct {
 	Props []*Prop
 }
 
+type DirectiveDefinition struct {
+	BaseFileInfo
+	Name       string
+	Args       []*Arg
+	Repeatable bool
+	Locations  []string
+}
+
+type DirectiveArg struct {
+	Name   string
+	Value  []string
+	IsList bool
+}
+
 type Directive struct {
-	string
+	Name          string
+	DirectiveArgs []*DirectiveArg
 }

--- a/lib/write.go
+++ b/lib/write.go
@@ -10,6 +10,7 @@ type MergedSchema struct {
 }
 
 func (ms *MergedSchema) StitchSchema(s *Schema) string {
+	numOfDirs := len(s.DirectiveDefinitions)
 	numOfQurs := len(s.Queries)
 	numOfMuts := len(s.Mutations)
 	numOfSubs := len(s.Subscriptions)
@@ -28,6 +29,41 @@ func (ms *MergedSchema) StitchSchema(s *Schema) string {
 		ms.buf.WriteString("subscription: Subscription\n")
 	}
 	ms.buf.WriteString("}\n\n")
+
+	if numOfDirs > 0 {
+		for _, q := range s.DirectiveDefinitions {
+			ms.buf.WriteString(`directive @`)
+			ms.buf.WriteString(q.Name)
+			if l := len(q.Args); l > 0 {
+				ms.buf.WriteString("(")
+				if l > 2 {
+					ms.buf.WriteString("\n")
+				}
+
+				for i, a := range q.Args {
+					ms.stitchArgument(a, l, i)
+				}
+
+				if l > 2 {
+					ms.buf.WriteString("\n")
+					ms.addIndent(1)
+				}
+				ms.buf.WriteString(")")
+			}
+			if q.Repeatable {
+				ms.buf.WriteString(" repeatable")
+			}
+			ms.buf.WriteString(" on ")
+			for i, a := range q.Locations {
+				if i != 0 {
+					ms.buf.WriteString(" | ")
+				}
+				ms.buf.WriteString(a)
+			}
+			ms.buf.WriteString("\n")
+		}
+		ms.buf.WriteString("\n\n")
+	}
 
 	if numOfQurs > 0 {
 		ms.buf.WriteString(`type Query {
@@ -66,9 +102,7 @@ func (ms *MergedSchema) StitchSchema(s *Schema) string {
 				ms.buf.WriteString("!")
 			}
 
-			if q.Directive != nil {
-				ms.buf.WriteString(" @" + q.Directive.string)
-			}
+			ms.stitchDirectives(q.Directives)
 
 			ms.buf.WriteString("\n")
 		}
@@ -112,9 +146,7 @@ func (ms *MergedSchema) StitchSchema(s *Schema) string {
 				ms.buf.WriteString("!")
 			}
 
-			if m.Directive != nil {
-				ms.buf.WriteString(" @" + m.Directive.string)
-			}
+			ms.stitchDirectives(m.Directives)
 
 			ms.buf.WriteString("\n")
 		}
@@ -158,9 +190,7 @@ func (ms *MergedSchema) StitchSchema(s *Schema) string {
 				ms.buf.WriteString("!")
 			}
 
-			if c.Directive != nil {
-				ms.buf.WriteString(" @" + c.Directive.string)
-			}
+			ms.stitchDirectives(c.Directives)
 
 			ms.buf.WriteString("\n")
 		}
@@ -208,9 +238,7 @@ func (ms *MergedSchema) StitchSchema(s *Schema) string {
 				ms.buf.WriteString("!")
 			}
 
-			if p.Directive != nil {
-				ms.buf.WriteString(" @" + p.Directive.string)
-			}
+			ms.stitchDirectives(p.Directives)
 
 			ms.buf.WriteString("\n")
 		}
@@ -222,7 +250,9 @@ func (ms *MergedSchema) StitchSchema(s *Schema) string {
 	ms.buf.WriteString("\n")
 
 	for i, c := range s.Scalars {
-		ms.buf.WriteString("scalar " + c.Name + "\n")
+		ms.buf.WriteString("scalar " + c.Name)
+		ms.stitchDirectives(c.Directives)
+		ms.buf.WriteString("\n")
 		if i != len(s.Scalars)-1 {
 			ms.buf.WriteString("\n")
 		}
@@ -230,10 +260,14 @@ func (ms *MergedSchema) StitchSchema(s *Schema) string {
 	ms.buf.WriteString("\n")
 
 	for i, e := range s.Enums {
-		ms.buf.WriteString("enum " + e.Name + " {\n")
-		for _, n := range e.Fields {
+		ms.buf.WriteString("enum " + e.Name)
+		ms.stitchDirectives(e.Directives)
+		ms.buf.WriteString(" {\n")
+		for _, n := range e.EnumValues {
 			ms.addIndent(1)
-			ms.buf.WriteString(n + "\n")
+			ms.buf.WriteString(n.Name)
+			ms.stitchDirectives(n.Directives)
+			ms.buf.WriteString("\n")
 		}
 		ms.buf.WriteString("}\n")
 		if i != len(s.Enums)-1 {
@@ -243,7 +277,9 @@ func (ms *MergedSchema) StitchSchema(s *Schema) string {
 	ms.buf.WriteString("\n")
 
 	for j, i := range s.Interfaces {
-		ms.buf.WriteString("interface " + i.Name + " {\n")
+		ms.buf.WriteString("interface " + i.Name)
+		ms.stitchDirectives(i.Directives)
+		ms.buf.WriteString(" {\n")
 
 		for _, p := range i.Props {
 			ms.addIndent(1)
@@ -279,9 +315,7 @@ func (ms *MergedSchema) StitchSchema(s *Schema) string {
 				ms.buf.WriteString("!")
 			}
 
-			if p.Directive != nil {
-				ms.buf.WriteString(" @" + p.Directive.string)
-			}
+			ms.stitchDirectives(p.Directives)
 
 			ms.buf.WriteString("\n")
 		}
@@ -293,7 +327,9 @@ func (ms *MergedSchema) StitchSchema(s *Schema) string {
 	ms.buf.WriteString("\n")
 
 	for _, u := range s.Unions {
-		ms.buf.WriteString("union " + u.Name + " = ")
+		ms.buf.WriteString("union " + u.Name)
+		ms.stitchDirectives(u.Directives)
+		ms.buf.WriteString(" = ")
 		fields := strings.Join(u.Fields, " | ")
 		ms.buf.WriteString(fields + "\n\n")
 	}
@@ -318,9 +354,7 @@ func (ms *MergedSchema) StitchSchema(s *Schema) string {
 				ms.buf.WriteString("!")
 			}
 
-			if p.Directive != nil {
-				ms.buf.WriteString(" @" + p.Directive.string)
-			}
+			ms.stitchDirectives(p.Directives)
 
 			ms.buf.WriteString("\n")
 		}
@@ -363,6 +397,53 @@ func (ms *MergedSchema) stitchArgument(a *Arg, l int, i int) {
 		}
 		if !a.Null {
 			ms.buf.WriteString("!")
+		}
+	}
+
+	if l <= 2 && i != l-1 {
+		ms.buf.WriteString(", ")
+	}
+	if l > 2 && i != l-1 {
+		ms.buf.WriteString("\n")
+	}
+}
+
+func (ms *MergedSchema) stitchDirectives(a []*Directive) {
+	if l := len(a); l > 0 {
+		for _, a := range a {
+			ms.buf.WriteString(" @" + a.Name)
+			if m := len(a.DirectiveArgs); m > 0 {
+				ms.buf.WriteString("(")
+				for i, b := range a.DirectiveArgs {
+					ms.stitchDirectiveArgument(b, m, i)
+				}
+				ms.buf.WriteString(")")
+			}
+		}
+	}
+}
+
+func (ms *MergedSchema) stitchDirectiveArgument(a *DirectiveArg, l int, i int) {
+	if l > 2 {
+		ms.addIndent(2)
+	}
+	ms.buf.WriteString(a.Name + ": ")
+
+	if a.IsList {
+		ms.buf.WriteString("[")
+		for i, v := range a.Value {
+			if i != 0 {
+				ms.buf.WriteString(",")
+			}
+			ms.buf.WriteString(v)
+		}
+		ms.buf.WriteString("]")
+	} else {
+		for i, v := range a.Value {
+			if i != 0 {
+				ms.buf.WriteString(",")
+			}
+			ms.buf.WriteString(v)
 		}
 	}
 

--- a/main.go
+++ b/main.go
@@ -5,8 +5,8 @@ import (
 	"io/ioutil"
 	"os"
 
-	"github.com/mattdamon108/gqlmerge/command"
-	gql "github.com/mattdamon108/gqlmerge/lib"
+	"github.com/mununki/gqlmerge/command"
+	gql "github.com/mununki/gqlmerge/lib"
 )
 
 func main() {

--- a/test/CheckIfExists.graphql
+++ b/test/CheckIfExists.graphql
@@ -1,0 +1,17 @@
+type Query {
+  checkIfExists(userId: ID!): CheckIfExistsResponse!
+}
+
+type CheckIfExistsResponse {
+  ok: Boolean!
+  error: String
+  user: [User]!
+}
+
+type User {
+  id: ID!
+  email: String!
+  fullName: String!
+}
+
+scalar Price

--- a/test/Directives.graphql
+++ b/test/Directives.graphql
@@ -1,0 +1,16 @@
+directive @goModel(
+  model: String
+  models: [String!]
+) on OBJECT | INPUT_OBJECT | SCALAR | ENUM | INTERFACE | UNION
+
+"""
+TEST
+"""
+interface Node @goModel(model: "todo/ent.Noder") {
+  id: ID! # TEST
+}
+
+enum Color @goModel(model: "backend/ent/color.Color") {
+  Blue @ignore(if: isError)
+  Red
+}

--- a/test/GetMyProfile.graphql
+++ b/test/GetMyProfile.graphql
@@ -1,0 +1,22 @@
+directive @goModel(
+  model: String
+  models: [String!]
+) on OBJECT | INPUT_OBJECT | SCALAR | ENUM | INTERFACE | UNION
+
+type Query {
+  getMyProfile: UserResponse!
+}
+
+type UserResponse {
+  ok: Boolean!
+  error: String
+  user: User
+}
+
+type User {
+  id: ID!
+  email: String!
+  fullName: String!
+}
+
+scalar Price

